// rclnodejs/generated/msg_interfaces/msg_interfaces__srv__StopRecord_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EmptyWrapper = require('../../generated/std_msgs/std_msgs__msg__Empty.js');
const StopRecord_RequestRefStruct = StructType({
  stop_record_command: EmptyWrapper.refObjectType,
});
const StopRecord_RequestRefArray = ArrayType(StopRecord_RequestRefStruct);
const StopRecord_RequestRefStructArray = StructType({
  data: StopRecord_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StopRecord_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StopRecord_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.stop_record_command =  new EmptyWrapper(other._wrapperFields.stop_record_command);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StopRecord_RequestRefStruct();
    this._wrapperFields.stop_record_command = new EmptyWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new StopRecord_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StopRecord_RequestArrayWrapper;
  }
  static get ArrayType() {
    return StopRecord_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return StopRecord_RequestRefStructArray
  }
  static get refObjectType() {
    return StopRecord_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.stop_record_command.freeze(own, checkConsistency);
    this._refObject.stop_record_command = this._wrapperFields.stop_record_command.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.stop_record_command.copyRefObject(refObject.stop_record_command);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    EmptyWrapper.freeStruct(refObject.stop_record_command);
  }
  static destoryRawROS(msg) {
    StopRecord_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'msg_interfaces', subFolder: 'srv', interfaceName: 'StopRecord_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get stop_record_command() {
    return this._wrapperFields.stop_record_command;
  }
  set stop_record_command(value) {
    if (value instanceof EmptyWrapper) {
      this._wrapperFields.stop_record_command.copy(value);
    } else {
      this._wrapperFields.stop_record_command.copy(new EmptyWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new StopRecord_RequestRefStruct(refObject.toObject());
    this._wrapperFields.stop_record_command.copyRefObject(this._refObject.stop_record_command);
  }
  copy(other) {
    this._refObject = new StopRecord_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.stop_record_command.copy(other._wrapperFields.stop_record_command);
  }
  static get classType() {
    return StopRecord_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"stop_record_command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Empty","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"msg_interfaces","type":"StopRecord_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StopRecord_Request"};
  }
  hasMember(name) {
    let memberNames = ["stop_record_command"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StopRecord_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StopRecord_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StopRecord_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StopRecord_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StopRecord_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StopRecord_RequestRefArray(size);
    this._refObject = new StopRecord_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StopRecord_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StopRecord_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StopRecord_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StopRecord_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StopRecord_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StopRecord_RequestArrayWrapper;
  }
}
module.exports = StopRecord_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "stop_record_command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Empty",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "msg_interfaces",
    "type": "StopRecord_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StopRecord_Request"
}
*/
